README.md
# michaeldmoore-annunciator-panel

## Custom annunciator panel for grafana, based on the built-in SingleStat panel

This panel was developed for as an enhancement to Singlestat providing some useful additions particularly suited to process control/monitoring dashboards.  Singlestat displays a single metric from time series data set, with optional threshold-related coloring etc.  Annunciator builds on this base, adding optional upper and lower warning, as well as hard limits.  If desired, the upper and lower limits can be displayed as reference values.  Each of the 5 possible states (OK, upper or lower limits exceeded and upper or lower warning levels exceeded) can be disabled.  if enabled, these can individually be set to 'flash' (using SVG animation).

![Showcase](https://raw.githubusercontent.com/michaeldmoore/michaeldmoore-annunciator-panel/master/src/img/Showcase.gif)

Like SingleStat, the Annunciator Panel allows you to show the one main summary stat of a SINGLE series. It reduces the series into a single number (by looking at the max, min, average, or sum of values in the series). Depending on the configuration options, the panel can also display upper and lower alarm thresholds, colored to according to the value of the metric relative to the thresholds, and optional flash etc.

### Features

In addition to the basic features of SingleStat, this panel provides additional display attributes including the ability to set upper and lower warning and limit thresholds, and to set the colors appropriate to each of these states and whether or not the values should 'flash'.

If desired, the displayed value can be set to flash when either of the limit values are exceeded.  The current values of these limits can also be displayed on the annunciator.

![Options](https://raw.githubusercontent.com/michaeldmoore/michaeldmoore-annunciator-panel/master/src/img/Options.png)

This panel is designed to support simple time series data sets - unlike SingleStat though, Annunciator does not support table-based data sets.

## Configuration details

The Options tab is made of three groups, **Value**, **Thresholds** and **Spark Lines**.

The **Value group** has a setting for the Metric - that is, the scalar operation applied to the data series.  Most commonly, this will be the default 'Current' value.

![Values-MetricNames](https://raw.githubusercontent.com/michaeldmoore/michaeldmoore-annunciator-panel/master/src/img/Values-MetricNames.png)

The corresponding Font size setting controls the relative size of the main displayed value.

![Values-MetricFontSize](https://raw.githubusercontent.com/michaeldmoore/michaeldmoore-annunciator-panel/master/src/img/Values-MetricFontSize.png)

The next two rows in this group provide for an optional prefix and/or postfix string to be displayed either side of the main value.  A special font size value 'hide' can be used to suppress the display of either of these strings.

![Values-PostfixFontSize](https://raw.githubusercontent.com/michaeldmoore/michaeldmoore-annunciator-panel/master/src/img/Values-PostfixFontSize.png)

The final part of this group sets the units of measurement.

![Values-Units](https://raw.githubusercontent.com/michaeldmoore/michaeldmoore-annunciator-panel/master/src/img/Values-Units.png)

The **Thresholds group** controls the various display enhancements.
In the default configuration, both the upper and lower limits are disabled.  The color, number of decimal places and font size can be configured independently for the single metric.

![Thresholds-NoLimits](https://raw.githubusercontent.com/michaeldmoore/michaeldmoore-annunciator-panel/master/src/img/Thresholds-NoLimits.png)

Note the rows for both Upper and Lower limits are set to 'disabled'.

Setting either or both of these to 'static', 'flash', or the new value - 'shock & awe' exposes yet more settings - the threshold value(s) for the limit(s) along with Color, number of decimal places and font size.  When limits are enabled, the corresponding warning thresholds also appear, here in the default 'disabled' state.

![Thresholds-UpperLimit](https://raw.githubusercontent.com/michaeldmoore/michaeldmoore-annunciator-panel/master/src/img/Thresholds-UpperLimit.png)

In this example, the main value display will be green whenever the value is &lt; 80\. and red when above 80\.  The limit itself will also appear above and to the right of the main value display.  The font size of this displayed limit indicator can be adjusted using the Font Size selection - or hidden, if desired, by using the special 'hide' font size setting.  Note too, that the number of decimal places for the limit display can be defined - or left in the default 'Auto' mode.  If the limit display option is set to 'flash', the main display value - and the corresponding limit display will both flash (once per second) whenever the threshold value is exceeded.

Once any of the limit thresholds are enabled, the corresponding warning thresholds may also be set as 'static' or 'flash', along with the color value.  Once enabled, this color - and possible flash behavior - will appear whenever the value exceeds the warning level but not the limit level.  Note that the warning levels themselves are not displayed on the panel, hence the number of decimal places and font size settings are not provided for warning thresholds.

![Thresholds-UpperWarning](https://raw.githubusercontent.com/michaeldmoore/michaeldmoore-annunciator-panel/master/src/img/Thresholds-UpperWarning.png)&nbsp;

Notice how the OK Value label is updated to show the new 'OK' range - in this case anything below 75.

With the full set of thresholds enabled, all the settings are now exposed.

![Thresholds-AllSettings](https://raw.githubusercontent.com/michaeldmoore/michaeldmoore-annunciator-panel/master/src/img/Thresholds-AllSettings.png)

The **Spark Lines group** is preserved unchanged from the original SingleStat panel.  See the [SingleStat documentation for
configuration details](http://docs.grafana.org/features/panels/singlestat/#spark-lines) pertaining to this feature.

![Options](https://raw.githubusercontent.com/michaeldmoore/michaeldmoore-annunciator-panel/master/src/img/SparkLines.png)


src/CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## v1.0.5

Previous Release

## v1.1.0

Signed, as required by Grafana 7.x.x.  Removed some internal dependencies, general clean-up

gruntfile.js
/*jshint esversion: 6 */
module.exports = (grunt) => {
  require('load-grunt-tasks')(grunt);

  grunt.loadNpmTasks('grunt-execute');
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-jshint');

  grunt.initConfig({

    clean: ['dist'],

    copy: {
      src_to_dist: {
        cwd: 'src',
        expand: true,
        src: ['**/*', '!**/*.js', '!**/*.scss', '!img/**/*'],
        dest: 'dist'
      },
      pluginDef: {
        expand: true,
        src: ['README.md'],
        dest: 'dist'
      },
      img_to_dist: {
        cwd: 'src',
        expand: true,
        src: ['img/michaeldmoore-annunciator-panel.svg', 'img/Options.png', 'img/Showcase.gif'],
        dest: 'dist'
      },
      css_to_dist: {
        cwd: 'src',
        expand: true,
        src: ['css/**/*'],
        dest: 'dist'
      }
    },

    jshint: {
      all: ['Gruntfile.js', 'src/*.js']
    },

    babel: {
      options: {
        sourceMap: true,
        presets: ['es2015'],
        plugins: ['transform-es2015-modules-systemjs', 'transform-es2015-for-of']
      },
      dist: {
        files: [{
          cwd: 'src',
          expand: true,
          src: ['*.js'],
          dest: 'dist',
          ext: '.js'
        }]
      }
    },

    watch: {
      rebuild_all: {
        files: ['src/**/*'],
        tasks: ['default'],
        options: { spawn: false }
      }
    }

  });

  grunt.registerTask('default', ['clean', 'jshint', 'copy:src_to_dist', 'copy:pluginDef', 'copy:img_to_dist', 'copy:css_to_dist', 'babel']);
};


src/module.js
/*jshint esversion: 6 */
/*jshint -W087 */
import {
  MetricsPanelCtrl
} from 'app/plugins/sdk';

import "./css/annunciator-panel.css!";

import _ from 'lodash';
import $ from 'jquery';
import 'jquery.flot';
import kbn from 'app/core/utils/kbn';
import TimeSeries from 'app/core/time_series2';
import { getTemplateSrv } from '@grafana/runtime';
import { PanelEvents } from "@grafana/data";

const templateSrv = getTemplateSrv();

function isNumber(n) {
  return !isNaN(parseFloat(n)) && isFinite(n);
}

class AnnunciatorPanelCtrl extends MetricsPanelCtrl {

  /** @ngInject */
  constructor($scope, $injector) {
    super($scope, $injector);

    var panelDefaults = {
      "LowerLimit": {
        "DisplayOption": "disabled",
        "Color": "rgb(2, 17, 249)",
        "Value": 20,
        "Decimals": 1,
        "FontSize": "50%"
      },
      "LowerWarning": {
        "DisplayOption": "disabled",
        "Color": "rgb(9, 115, 102)",
        "Value": 25
      },
      "Metric": {
        "Name": "current",
        "Format": "percent",
        "Color": "rgb(2, 247, 2)",
        "Decimals": 4,
        "FontSize": "100%"
      },
      "UpperWarning": {
        "DisplayOption": "disabled",
        "Color": "rgb(247, 2, 2)",
        "Value": 75
      },
      "UpperLimit": {
        "DisplayOption": "disabled",
        "Color": "rgb(247, 90, 7)",
        "Value": 80,
        "Decimals": 2,
        "FontSize": "50%"
      },
      "Prefix": {
        "Text": "",
        "FontSize": "hide",
      },
      "Postfix": {
        "Text": "",
        "FontSize": "hide",
      },
      "sparkline": {
        "fillColor": "rgba(19, 193, 91, 0.32)",
        "full": false,
        "lineColor": "rgb(145, 200, 16)",
        "show": true
      }
    };

    _.defaults(this.panel, panelDefaults);

    this.events.on(
      PanelEvents.dataReceived,
      this.onDataReceived.bind(this),
      $scope
    );

    this.events.on(
      PanelEvents.dataError,
      this.onDataError.bind(this),
      $scope
    );

    this.events.on(
      PanelEvents.render,
      this.onRender.bind(this)
    );

    this.events.on(
      PanelEvents.dataSnapshotLoad,
      this.onDataSnapshotLoad.bind(this)
    );

    this.events.on(
      PanelEvents.editModeInitialized,
      this.onInitEditMode.bind(this)
    );
  }

  onDataError(err) {
    this.errorMessage = 'Query Failure : ' + err.status + ', ' + err.statusText;
    this.seriesList = [];
    this.render([]);
  }

  onDataSnapshotLoad(snapshotData) {
    this.onDataReceived(snapshotData);
  }

  onInitEditMode() {
    this.metricNames = ['min', 'max', 'avg', 'current', 'total', 'first', 'delta', 'diff', 'range'];
    this.fontSizes = ['20%', '30%', '50%', '70%', '80%', '100%', '110%', '120%', '150%', '170%',
      '200%', '230%', '260%', '300%', '340%', '380%', '440%', '500%', '600%', '800%', '1000%'];
    this.fontSizes0 = ['hide'].concat(this.fontSizes);
    this.displayStates = ['disabled', 'static'];
    this.displayStates1 = this.displayStates.concat(['flash', 'shock & awe']);
    this.unitFormats = kbn.getUnitFormats();
    this.addEditorTab('Options', 'public/plugins/michaeldmoore-annunciator-panel/options.html', 2);
  }

  setUnitFormat(subItem) {
    this.panel.Metric.Format = subItem.value;
    this.render();
  }

  getValueRegion(value) {
    var valueRegion = "OK";

    if (this.panel.UpperLimit.DisplayOption != 'disabled') {
      if (value >= this.panel.UpperLimit.Value)
        valueRegion = "UpperLimit";
      else if (this.panel.UpperWarning.DisplayOption != 'disabled' && (value >= this.panel.UpperWarning.Value))
        valueRegion = "UpperWarning";
    }

    if (this.panel.LowerLimit.DisplayOption != 'disabled') {
      if (value <= this.panel.LowerLimit.Value)
        valueRegion = "LowerLimit";
      else if (this.panel.LowerWarning.DisplayOption != 'disabled' && (value <= this.panel.LowerWarning.Value))
        valueRegion = "LowerWarning";
    }

    return valueRegion;
  }

  getDisplayAttributesForValue(value, metric) {
    var displayAttributes = {};
    var color = this.panel.Metric.Color;
    var displayOption = this.panel.Metric.DisplayOption;
    var flash = false;

    var valueRegion = this.getValueRegion(this.data.value);
    switch (valueRegion) {
      case "UpperLimit":
        color = this.panel.UpperLimit.Color;
        displayOption = this.panel.UpperLimit.DisplayOption;
        break;

      case "UpperWarning":
        color = this.panel.UpperWarning.Color;
        displayOption = this.panel.UpperWarning.DisplayOption;
        break;

      case "LowerWarning":
        color = this.panel.LowerWarning.Color;
        displayOption = this.panel.LowerWarning.DisplayOption;
        break;

      case "LowerLimit":
        color = this.panel.LowerLimit.Color;
        displayOption = this.panel.LowerLimit.DisplayOption;
        break;
    }

    switch (metric) {
      case "UpperLimit":
        color = this.panel.UpperLimit.Color;
        if (displayOption == "shock & awe")
          displayOption = "flash";
        if (valueRegion != "UpperLimit")
          displayOption = "static";
        break;

      case "LowerLimit":
        color = this.panel.LowerLimit.Color;
        if (displayOption == "shock & awe")
          displayOption = "flash";
        if (valueRegion != "LowerLimit")
          displayOption = "static";
        break;
    }

    displayAttributes.color = color;
    displayAttributes.displayOption = displayOption;

    return displayAttributes;
  }

  buildValueHtml() {
    var html = '';

    html += "<div class='michaeldmoore-annunciator-panel-value-container'>";
    if (this.panel.Prefix.Text != '' && this.panel.Prefix.FontSize != 'hide')
      html += this.getTextSpan('michaeldmoore-annunciator-panel-prefix', this.panel.Prefix.FontSize, this.panel.Prefix.Text, {});

    html += this.getValueSpan('michaeldmoore-annunciator-panel-value', this.panel.Metric.FontSize, this.panel.Metric.Decimals, this.data.value, "Value");

    if (this.panel.Postfix.Text != '' && this.panel.Postfix.FontSize != 'hide')
      html += this.getTextSpan('michaeldmoore-annunciator-panel-postfix', this.panel.Postfix.FontSize, this.panel.Postfix.Text, {});

    html += "</div>";

    return html;
  }

  buildLimitsHtml() {
    var html = '';

    if (this.panel.UpperLimit.DisplayOption != 'disabled' || this.panel.LowerLimit.DisplayOption != 'disabled') {
      html += "<div class='michaeldmoore-annunciator-panel-limits-container'>";

      if (this.panel.LowerLimit.DisplayOption != 'disabled' && (this.panel.LowerLimit.FontSize != 'hide'))
        html += this.getValueSpan('michaeldmoore-annunciator-panel-lower-limit', this.panel.LowerLimit.FontSize, this.panel.LowerLimit.Decimals, this.panel.LowerLimit.Value, "LowerLimit");

      if (this.panel.UpperLimit.DisplayOption != 'disabled' && (this.panel.UpperLimit.FontSize != 'hide'))
        html += this.getValueSpan('michaeldmoore-annunciator-panel-upper-limit', this.panel.UpperLimit.FontSize, this.panel.UpperLimit.Decimals, this.panel.UpperLimit.Value, "UpperLimit");

      html += "</div>";

      html += "<div style='float:none;'/>";
    }
    return html;
  }

  buildHtml() {
    var html = "<div class='michaeldmoore-annunciator-panel-container' style='height:100%;'>";

    if (this.data == null || this.data.value == null)
      this.errorMessage = 'No data';

    if (this.errorMessage) {
      html += '<div class="michaeldmoore-annunciator-panel-centered">' + this.errorMessage + '</div>';
    }
    else {
      html += this.buildLimitsHtml();
      if (isNumber(this.data.value))
        html += this.buildValueHtml();
    }

    html += "</div>";

    this.elem.html(html);

    if (this.panel.sparkline.show) {
      this.addSparkline();
    }
  }

  getTextSpan(className, fontSize, text, displayAttributes) {
    var style = '';

    if (fontSize)
      style += "font-size:" + fontSize + ";";

    if (displayAttributes && displayAttributes.color)
      style += "color:" + displayAttributes.color + ";";

    if (style)
      style = ' style="' + style + '"';

    if (displayAttributes && displayAttributes.displayOption)
      className += " michaeldmoore-annunciator-" + displayAttributes.displayOption.replace(' & ', 'and');

    return '<span class="' + className + '"' + style + '>' + text + '</span>';
  }

  getValueSpan(className, fontSize, decimals, value, metric) {
    var displayAttributes = this.getDisplayAttributesForValue(value, metric);
    return this.getTextSpan(className, fontSize, this.formatValue(value, decimals), displayAttributes);
  }

  setOKValueRange() {
    var OKLowerLimit;
    if (this.panel.LowerLimit.DisplayOption != 'disabled') {
      if (isNumber(this.panel.LowerLimit.Value)) {
        if (this.panel.LowerWarning.DisplayOption != 'disabled') {
          if (isNumber(this.panel.LowerWarning.Value))
            if (Number(this.panel.LowerWarning.Value) > Number(this.panel.LowerLimit.Value))
              OKLowerLimit = this.panel.LowerWarning.Value;
            else
              this.errorMessage = 'Warning : LowerWarning must be greater than LowerLimit';
          else
            this.errorMessage = 'Warning : LowerWarning is non-numeric';
        } else
          OKLowerLimit = this.panel.LowerLimit.Value;
      } else {
        this.errorMessage = 'Warning : LowerLimit is non-numeric';
      }
    }

    var OKUpperLimit;
    if (this.panel.UpperLimit.DisplayOption != 'disabled') {
      if (isNumber(this.panel.UpperLimit.Value)) {
        if (this.panel.UpperWarning.DisplayOption != 'disabled') {
          if (isNumber(this.panel.UpperWarning.Value))
            if (Number(this.panel.UpperWarning.Value) < Number(this.panel.UpperLimit.Value))
              OKUpperLimit = this.panel.UpperWarning.Value;
            else
              this.errorMessage = 'Warning : UpperWarning must be less than UpperLimit';
          else
            this.errorMessage = 'Warning : UpperWarning is non-numeric';
        } else
          OKUpperLimit = this.panel.UpperLimit.Value;
      } else {
        this.errorMessage = 'Warning : UpperLimit is non-numeric';
      }
    }

    if (isNumber(OKLowerLimit) && isNumber(OKUpperLimit)) {
      if (OKUpperLimit > OKLowerLimit) {
        this.panel.MetricValueRange = OKLowerLimit + ' -> ' + OKUpperLimit;
      } 
      else {
        this.errorMessage = 'Warning : UpperWarning/limit must be more than LowerWarning/Limit';
        this.panel.MetricValueRange = '?????';
      }
    }
    else if (isNumber(OKLowerLimit))
      this.panel.MetricValueRange = '> ' + OKLowerLimit;
    else if (isNumber(OKUpperLimit))
      this.panel.MetricValueRange = '< ' + OKUpperLimit;
    else
      this.panel.MetricValueRange = '';
  }

  onRender() {
    var data = {};
    if (this.dataList.length > 0) {
      this.dataList[0].datapoints = this.dataPoints.filter(dp => dp[1] >= this.range.from && dp[1] <= this.range.to);
      if (this.dataList[0].datapoints.length > 0) {
        this.series = this.dataList.map(this.seriesHandler.bind(this));
        this.setValues(data);
      }
    }

    this.data = data;

    this.setOKValueRange();
    this.buildHtml();
    this.ctrl.renderingCompleted();
    delete this.errorMessage;
  }


  onDataReceived(dataList) {
    this.dataList = dataList;
    if (dataList.length > 0) {
      this.dataPoints = dataList[0].datapoints;
    }
    this.render();
  }

  seriesHandler(seriesData) {
    var series = new TimeSeries({
      datapoints: seriesData.datapoints,
      alias: seriesData.target,
    });

    series.flotpairs = series.getFlotPairs(this.panel.nullPointMode);
    return series;
  }

  getDecimalsForValue(value, decimals) {
    if (_.isNumber(decimals)) {
      return {
        decimals: decimals,
        scaledDecimals: null
      };
    }

    var delta = value / 2;
    var dec = -Math.floor(Math.log(delta) / Math.LN10);

    var magn = Math.pow(10, -dec),
      norm = delta / magn, // norm is between 1.0 and 10.0
      size;

    if (norm < 1.5) {
      size = 1;
    } else if (norm < 3) {
      size = 2;
      // special case for 2.5, requires an extra decimal
      if (norm > 2.25) {
        size = 2.5;
        ++dec;
      }
    } else if (norm < 7.5) {
      size = 5;
    } else {
      size = 10;
    }

    size *= magn;

    // reduce starting decimals if not needed
    if (Math.floor(value) === value) {
      dec = 0;
    }

    var result = {};
    result.decimals = Math.max(0, dec);
    result.scaledDecimals = result.decimals - Math.floor(Math.log(size) / Math.LN10) + 2;

    return result;
  }

  formatValue(value, decimals) {
    // crude work-around for kbn formatting function error - preserve decimal places even for whole numbers
    //if (value == 0 && decimals > 0)
    //    value += 0.000000000000001;
    var decimalInfo = this.getDecimalsForValue(value, decimals);
    var formatFunc = kbn.valueFormats[this.panel.Metric.Format];
    return formatFunc(value, decimalInfo.decimals, decimalInfo.scaledDecimals);
  }

  setValues(data) {
    data.flotpairs = [];

    //if (this.series.length > 1) {
    //  this.errorMessage = 'Multiple Series Error : query returns ' + this.series.length + ' series. Panel expects a single series';
    //}

    if (this.series && this.series.length > 0) {
      var lastPoint = _.last(this.series[0].datapoints);
      var lastValue = _.isArray(lastPoint) ? lastPoint[0] : null;

      if (_.isString(lastValue)) {
        data.value = 0;
        data.valueFormatted = _.escape(lastValue);
        data.valueRounded = 0;
      } else {
        data.value = this.series[0].stats[this.panel.Metric.Name];
        data.flotpairs = this.series[0].flotpairs;

        if (data == null || data.value == null) {
          data.value = 0.0;
        }

        var decimalInfo = this.getDecimalsForValue(data.value, this.panel.Metric.Decimals);
        var formatFunc = kbn.valueFormats[this.panel.Metric.Format];
        data.valueFormatted = formatFunc(data.value, decimalInfo.decimals, decimalInfo.scaledDecimals);
        data.valueRounded = kbn.roundValue(data.value, decimalInfo.decimals);
      }

      if (data == null || data.value == null) {
        data.value = 0.0;
      }
    }
  }

  addSparkline() {
    if (this.elem == null || this.elem.width() < 10) {
      // element has not gotten it's width yet
      // delay sparkline render
      setTimeout(this.addSparkline, 30);
      return;
    }

    var width = this.elem.width() + 20;
    var height = this.elem.height();
    var plotCanvas = $('<div></div>');
    var plotCss = {};
    plotCss.position = 'absolute';

    if (this.panel.sparkline.full) {
      plotCss.bottom = '5px';
      plotCss.left = '-5px';
      plotCss.width = (width - 10) + 'px';
      var dynamicHeightMargin = height <= 100 ? 5 : (Math.round((height / 100)) * 15) + 5;
      plotCss.height = (height - dynamicHeightMargin) + 'px';
    } else {
      plotCss.bottom = "0px";
      plotCss.left = "-5px";
      plotCss.width = (width - 10) + 'px';
      plotCss.height = Math.floor(height * 0.25) + 'px';
    }

    plotCanvas.css(plotCss);

    var options = {
      legend: {
        show: false
      },
      series: {
        lines: {
          show: true,
          fill: 1,
          lineWidth: 1,
          fillColor: this.panel.sparkline.fillColor,
        },
      },
      yaxes: {
        show: false
      },
      xaxis: {
        show: false,
        mode: "time",
        min: this.ctrl.range.from.valueOf(),
        max: this.ctrl.range.to.valueOf(),
      },
      grid: {
        hoverable: false,
        show: false
      },
    };

    this.elem.find('.michaeldmoore-annunciator-panel-container').append(plotCanvas);

    var plotSeries = {
      data: this.data.flotpairs,
      color: this.panel.sparkline.lineColor
    };

    $.plot(plotCanvas, [plotSeries], options);
  }

  onConfigChanged() {
    this.refresh();
  }

  link(scope, elem, attrs, ctrl) {
    this.ctrl = ctrl;
    this.elem = elem;

    // for backward compatability (grafana 6.6.0 and earlier)
    var panelContentElem = elem.find('.panel-content');
    if (panelContentElem.length)
      this.elem = panelContentElem;
  }
}

AnnunciatorPanelCtrl.templateUrl = 'module.html';

export {
  AnnunciatorPanelCtrl as PanelCtrl
};